<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ondisc">
<title>Getting started with ondisc • ondisc</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with ondisc">
<meta property="og:description" content="ondisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ondisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/ondisc.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/timothy-barry/ondisc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with ondisc</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/timothy-barry/ondisc/blob/HEAD/vignettes/ondisc.Rmd" class="external-link"><code>vignettes/ondisc.Rmd</code></a></small>
      <div class="d-none name"><code>ondisc.Rmd</code></div>
    </div>

    
    
<p><a href="https://timothy-barry.github.io/ondisc/" class="external-link"><code>ondisc</code></a>
is a companion R package to <code>sceptre</code> that facilitates
analysis of large-scale single-cell data out-of-core on a laptop or
distributed across tens to hundreds processors on a cluster or cloud. In
both of these settings, <code>ondisc</code> requires only a few
gigabytes of memory, even if the input data are tens of gigabytes in
size. <code>ondisc</code> mainly is oriented toward single-cell CRISPR
screen analysis, but <code>ondisc</code> also can be used for
single-cell differential expression and single-cell co-expression
analyses. <code>ondisc</code> is powered by several new, efficient
algorithms for manipulating and querying large, sparse expression
matrices. Although <code>ondisc</code> and <code>sceptre</code> work
best in conjunction, <code>ondisc</code> can be used independently of
<code>sceptre</code> (and conversely, <code>sceptre</code> can be used
independently of <code>ondisc</code>).</p>
<p>Users can install <code>ondisc</code> using the code below.
<code>ondisc</code> depends on the Bioconductor package
<code>Rhdf5lib</code>, which should be installed from source before
installing <code>ondisc</code>. Users also should install
<code>sceptredata</code>, which contains the example data used in this
vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("BiocManager"); install.packages("devtools")</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Rhdf5lib"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span> <span class="co"># Rhdf5lib</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"timothy-barry/ondisc"</span><span class="op">)</span> <span class="co"># ondisc</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"katsevich-lab/sceptredata"</span><span class="op">)</span> <span class="co"># sceptredata</span></span></code></pre></div>
<p>See the <a href="https://timothy-barry.github.io/sceptre-book/faq.html" class="external-link">frequently
asked questions page</a> for tips on installing <code>ondisc</code> such
that it runs as fast as possible. We can load <code>ondisc</code> and
<code>sceptredata</code> by calling <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/ondisc/" class="external-link">ondisc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/sceptre-book/" class="external-link">sceptredata</a></span><span class="op">)</span></span></code></pre></div>
<p>The interface to <code>ondisc</code> is simple and minimal. The
package contains only one class: <code>odm</code> (short for
“<code>ondisc</code> matrix”). An <code>odm</code> object represents a
single-cell expression matrix stored <em>on disk</em> (as opposed to
<em>in memory</em>). <code>odm</code> objects can be used to store
expression matrices that are too large to fit in memory. Users can
create an <code>odm</code> object via one of two functions:
<code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> or
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>. The former takes the output of
one or more calls to Cell Ranger count as input, while the latter takes
an R matrix (stored in standard format or sparse format) as input. Users
can interface with an <code>odm</code> object using several functions,
including the bracket (<code>[,]</code>) operator, which loads a
specified subset of the expression matrix into memory.</p>
<div class="section level2">
<h2 id="initializing-an-odm-object-via-create_odm_from_cellranger">Initializing an <code>odm</code> object via
<code>create_odm_from_cellranger()</code><a class="anchor" aria-label="anchor" href="#initializing-an-odm-object-via-create_odm_from_cellranger"></a>
</h2>
<p><code>ondisc</code> provides two functions for initializing an
<code>odm</code> object: <code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> and
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>. The former is considerably more
scalable and memory-efficient than the latter; thus, we recommend that
users employ <code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> when possible. We
illustrate use of <code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> on an
example single-cell CRISPR screen dataset stored in the
<code>sceptredata</code> package. The example data contain two
modalities, namely a gene modality and a CRISPR gRNA modality. There are
526 genes, 95 gRNAs, and 45,919 cells in the data. Users can read more
about the example data by evaluating
<code><a href="https://bioconductor.org/packages/release/bioc/vignettes/sceptredata/inst/doc/sceptredata.html" class="external-link">vignette("sceptredata")</a></code> or
<code><a href="https://rdrr.io/pkg/sceptredata/man/highmoi_example_data.html" class="external-link">?highmoi_example_data</a></code> in the console.
<code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> takes several arguments:
<code>directories_to_load</code>, <code>directory_to_write</code>,
<code>write_cellwise_covariates</code>, <code>chunk_size</code>,
<code>compression_level</code>, and <code>grna_target_data_frame</code>.
Only the first two of these arguments are required; the rest are set to
reasonable defaults. We describe the <code>directories_to_load</code>
and <code>directory_to_write</code> arguments below.</p>
<p><code>directories_to_load</code> is a character vector specifying the
locations of one or more directories outputted by Cell Ranger count.
Below, we set <code>directories_to_load</code> to the (machine-specific)
location of the example data on disk.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">directories_to_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptredata"</span><span class="op">)</span>, </span>
<span>  <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">directories_to_load</span> <span class="co"># file paths to the example data on your computer</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/sceptredata/extdata/highmoi_example/gem_group_1"</span></span>
<span><span class="co">## [2] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/sceptredata/extdata/highmoi_example/gem_group_2"</span></span></code></pre>
<p><code>directories_to_load</code> contains the file paths to two
directories, which correspond to cells sequenced across two batches. The
data are stored in <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-mex-matrices" class="external-link">feature
barcode format</a>; each directory contains the files
<code>barcodes.tsv.gz</code>, <code>features.tsv.gz</code>, and
<code>matrix.mtx.gz</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "barcodes.tsv.gz" "features.tsv"    "matrix.mtx"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"</span></span></code></pre>
<p>Next, <code>directory_to_write</code> is a file path to the directory
in which to write the backing <code>.odm</code> file, which is the file
that will store the expression data on disk. <code>.odm</code> files
contain the same information as <code>.mtx</code> files but stored in a
more efficient format for CRISPR screen analysis, differential
expression analysis, and gene co-expression analysis. <code>.odm</code>
files simply are HDF5 files with special structure. We set
<code>directory_to_write</code> to <code>temp_dir</code> (i.e., the
temporary directory) in this example. The remaining arguments are
optional, and most users will not need to specify them; see
<code>?create_odm_from_cellranger()</code> for more information. Below,
we call <code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> on the example data,
saving the output of the function to the variable
<code>out_list</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories_to_load <span class="op">=</span> <span class="va">directories_to_load</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span> </span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>out_list</code> contains three entries: <code>gene</code>,
<code>grna</code>, and <code>cellwise_covariates</code>.
<code>gene</code> and <code>grna</code> are the <code>odm</code> objects
corresponding to the gene and gRNA modalities, respectively. Meanwhile,
<code>cellwise_covariates</code> is a data frame that contains the
cell-wise covariates. (More on the cell-wise covariates later.) An
inspection of <code>temp_dir</code> reveals that the files
<code>gene.odm</code> and <code>grna.odm</code> have been written to
this directory.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">temp_dir</span>, pattern <span class="op">=</span> <span class="st">"*.odm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene.odm" "grna.odm"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="interacting-with-the-odm-object">Interacting with the <code>odm</code> object<a class="anchor" aria-label="anchor" href="#interacting-with-the-odm-object"></a>
</h2>
<p>We extract the <code>odm</code> object corresponding to the gene
modality as follows.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="va">out_list</span><span class="op">[[</span><span class="st">"gene"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Evaluating an <code>odm</code> object in the console prints
information about the matrix, including the number of features and cells
contained within the matrix, as well as the file path to the
(machine-specific) backing <code>.odm</code> file.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_odm</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class <span style="color: #0000BB;">odm</span> with the following attributes:</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">526</span> features</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">45919</span> cells</span></span>
<span><span class="co">##  • Backing file: <span style="color: #0000BB;">/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/gene.odm</span></span></span></code></pre>
<p><code>odm</code> objects support several key matrix operations,
including <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol()</a></code>, <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code>,
<code><a href="../reference/rownames-odm-method.html">rownames()</a></code>, and <code>[,]</code>. <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol()</a></code> and
<code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code> return the number of rows (i.e., features) and
columns (i.e., cells) contained within the matrix, respectively.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="va">n_features</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 526</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="va">n_cells</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 45919</span></span></code></pre>
<p>Next, <code><a href="../reference/rownames-odm-method.html">rownames()</a></code> returns the feature IDs.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rownames-odm-method.html">rownames</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">feature_ids</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ENSG00000069275" "ENSG00000117222" "ENSG00000117266" "ENSG00000117280"</span></span>
<span><span class="co">## [5] "ENSG00000133059" "ENSG00000133065"</span></span></code></pre>
<p>Finally, the bracket operator (<code>[,]</code>) loads a specified
row of the expression matrix into memory. One can index into the rows by
integer index or feature ID, as follows.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">expression_vector</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2 1 0 1 1 0</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expression_vector</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="st">"ENSG00000117222"</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">expression_vector</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2 1 0 1 1 0</span></span></code></pre>
<p>Indexing into an <code>odm</code> object by column is not supported.
Finally, <code>odm</code> objects take up very little space, as the data
are stored on-disk rather than in-memory. For example,
<code>gene_odm</code> takes up only 40 kilobytes of memory.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span>units <span class="op">=</span> <span class="st">"Kb"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "38.7 Kb"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="supported-modalities">Supported modalities<a class="anchor" aria-label="anchor" href="#supported-modalities"></a>
</h2>
<p><code>ondisc</code> supports the following Cell Ranger modalities:
<code>Gene Expression</code>, <code>CRISPR Guide Capture</code> (i.e.,
gRNA expression), and <code>Antibody Capture</code> (i.e., protein
expression). (The modality of a given feature is listed within the third
column of the unzipped <code>features.tsv</code> file; see the <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-mex-matrices" class="external-link">Cell
Ranger documentation</a> for more information.) The table below maps the
modality name used by Cell Ranger to that used by
<code>ondisc</code>.</p>
<table class="table">
<thead><tr class="header">
<th>Cell Ranger modality name</th>
<th>
<code>ondisc</code> modality name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>Gene Expression</code></td>
<td><code>gene</code></td>
</tr>
<tr class="even">
<td><code>CRISPR Guide Capture</code></td>
<td><code>grna</code></td>
</tr>
<tr class="odd">
<td><code>Antibody Capture</code></td>
<td><code>protein</code></td>
</tr>
</tbody>
</table>
<p>We provide an example of using
<code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> to import a dataset containing
three modalities: gene expression, gRNA expression, and protein
expression. We use a synthetic dataset for this purpose (so as to reduce
the amount of data stored within the <code>sceptredata</code> package).
To this end we call the function
<code><a href="../reference/write_example_cellranger_dataset.html">write_example_cellranger_dataset()</a></code>, which creates a
synthetic single-cell dataset, writing the dataset to disk in Cell
Ranger feature barcode format. (See
<code>?write_example_cellranger_dataset()</code> for more information
about this function.) We create a synthetic single-cell dataset
consisting of 500 genes, 50 gRNAs, 20 proteins, and 10,000 cells.
Furthermore, we specify that the cells are sequenced across three
batches. We write the synthetic dataset to the directory
<code>temp_dir</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/write_example_cellranger_dataset.html">write_example_cellranger_dataset</a></span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">50</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>  n_cells <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  n_batch <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  modalities <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"grna"</span>, <span class="st">"protein"</span><span class="op">)</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span> ,</span>
<span>  p_set_col_zero <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The synthetic data are contained in the directories
<code>batch_1</code>, <code>batch_2</code>, and <code>batch_3</code>
within <code>temp_dir</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">directories_to_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  <span class="va">temp_dir</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"batch_"</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">directories_to_load</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/batch_1"</span></span>
<span><span class="co">## [2] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/batch_2"</span></span>
<span><span class="co">## [3] "/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/batch_3"</span></span></code></pre>
<p>Each of these directories contains the files
<code>matrix.mtx.gz</code>, <code>features.tsv.gz</code>, and
<code>barcodes.tsv.gz</code>. For example, the contents of the
<code>batch_1</code> are as follows.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">directories_to_load</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"</span></span></code></pre>
<p>We call <code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> to import these
data, saving the output of the function in the variable
<code>out_list</code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories_to_load <span class="op">=</span> <span class="va">directories_to_load</span>,</span>
<span>  directory_to_write <span class="op">=</span> <span class="va">temp_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>out_list</code> contains the cell-wise covariate data frame
alongside <code>odm</code> objects corresponding to the gene, gRNA, and
protein modalities.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene"                "grna"                "protein"            </span></span>
<span><span class="co">## [4] "cellwise_covariates"</span></span></code></pre>
<p>Moreover, the files <code>gene.odm</code>, <code>grna.odm</code>, and
<code>protein.odm</code> have been written to disk. (The previous
<code>gene.odm</code> and <code>grna.odm</code> files are
overwritten.)</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">temp_dir</span>, pattern <span class="op">=</span> <span class="st">"*.odm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene.odm"    "grna.odm"    "protein.odm"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="the-cell-wise-covariate-data-frame">The cell-wise covariate data frame<a class="anchor" aria-label="anchor" href="#the-cell-wise-covariate-data-frame"></a>
</h2>
<p>As part of importing the data,
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> computes the cell-wise
covariates. We print the first few rows of the cell-wise covariate data
frame corresponding to the synthetic data below.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellwise_covariates</span> <span class="op">&lt;-</span> <span class="va">out_list</span><span class="op">[[</span><span class="st">"cellwise_covariates"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cellwise_covariates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    gene_n_umis gene_n_nonzero gene_p_mito grna_n_umis grna_n_nonzero</span></span>
<span><span class="co">##          &lt;int&gt;          &lt;int&gt;       &lt;num&gt;       &lt;int&gt;          &lt;int&gt;</span></span>
<span><span class="co">## 1:        1030            196   0.4330097         131             22</span></span>
<span><span class="co">## 2:        1034            187   0.4197292         126             24</span></span>
<span><span class="co">## 3:        1142            203   0.4168126         126             20</span></span>
<span><span class="co">## 4:        1177            217   0.4188615         119             21</span></span>
<span><span class="co">## 5:        1083            207   0.4524469         118             24</span></span>
<span><span class="co">## 6:        1095            193   0.4000000          89             17</span></span>
<span><span class="co">##    grna_feature_w_max_expression grna_frac_umis_max_feature protein_n_umis</span></span>
<span><span class="co">##                           &lt;char&gt;                      &lt;num&gt;          &lt;int&gt;</span></span>
<span><span class="co">## 1:                       grna_33                 0.07633588             64</span></span>
<span><span class="co">## 2:                        grna_6                 0.07936508             46</span></span>
<span><span class="co">## 3:                       grna_31                 0.07936508             51</span></span>
<span><span class="co">## 4:                       grna_22                 0.08403361             44</span></span>
<span><span class="co">## 5:                       grna_20                 0.08474576             31</span></span>
<span><span class="co">## 6:                       grna_11                 0.11235955             56</span></span>
<span><span class="co">##    protein_n_nonzero   batch</span></span>
<span><span class="co">##                &lt;int&gt;  &lt;fctr&gt;</span></span>
<span><span class="co">## 1:                 9 batch_1</span></span>
<span><span class="co">## 2:                 9 batch_1</span></span>
<span><span class="co">## 3:                 8 batch_1</span></span>
<span><span class="co">## 4:                 8 batch_1</span></span>
<span><span class="co">## 5:                 7 batch_1</span></span>
<span><span class="co">## 6:                 9 batch_1</span></span></code></pre>
<p>The modality to which a given covariate corresponds (“gene”, “grna”,
or “protein”) is prepended to the name of the covariate. We describe
each covariate below.</p>
<ul>
<li><p><code>gene_n_umis</code>: the number of gene UMIs sequenced in a
given cell.</p></li>
<li><p><code>gene_n_nonzero</code>: the number of genes that exhibit
nonzero expression in a given cell.</p></li>
<li><p><code>gene_p_mito</code>: the fraction of gene transcripts that
map to mitochondrial genes in a given cell. (Mitochondrial genes are
identified as genes whose name starts with <code>"MT-"</code> or
<code>"mt-"</code>.)</p></li>
<li><p><code>grna_n_umis</code>: similar to <code>gene_n_umis</code> but
for the gRNA modality.</p></li>
<li><p><code>grna_n_nonzero</code>: similar to
<code>gene_n_nonzero</code> but for the gRNA modality.</p></li>
<li><p><code>grna_feature_w_max_expression</code>: the ID of the gRNA
that exhibits the maximum UMI count in a given cell.</p></li>
<li><p><code>grna_frac_umis_max_feature</code>: the fraction of UMIs
that the maximally expressed gRNA in a given cell constitutes.</p></li>
<li><p><code>protein_n_umis</code>: similar to <code>gene_n_umis</code>
but for the protein modality.</p></li>
<li><p><code>protein_n_nonzero</code>: similar to
<code>gene_n_nonzero</code> but for the protein modality.</p></li>
<li><p><code>batch</code>: the batch in which a given cell was
sequenced. Cells loaded from different directories are assumed to belong
to different batches.</p></li>
</ul>
<p><code>sceptre</code> uses the covariates
<code>grna_feature_w_max_expression</code> and
<code>grna_frac_umis_max_feature</code> to assign gRNAs to cells.</p>
</div>
<div class="section level2">
<h2 id="reading-an--odm-file-into-r">Reading an <code>.odm</code> file into R<a class="anchor" aria-label="anchor" href="#reading-an--odm-file-into-r"></a>
</h2>
<p>Users can read an <code>.odm</code> file into R by calling the
function <code><a href="../reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file()</a></code>. Below, we
delete all variables from the global namespace. Then, we call
<code><a href="../reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file()</a></code> on the file
<code>gene.odm</code> stored within <code>temp_dir</code>, which loads
the gene expression matrix that we created in the previous step.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># delete all variables</span></span>
<span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initialize_odm_from_backing_file.html">initialize_odm_from_backing_file</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"/gene.odm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gene_odm</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class <span style="color: #0000BB;">odm</span> with the following attributes:</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">500</span> features</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">10000</span> cells</span></span>
<span><span class="co">##  • Backing file: <span style="color: #0000BB;">/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/gene.odm</span></span></span></code></pre>
<p><code>.odm</code> files are portable. Thus, a user can create an
<code>.odm</code> file on one computer, move the <code>.odm</code> file
to another computer, and then open the <code>.odm</code> file on the
second computer.</p>
</div>
<div class="section level2">
<h2 id="initializing-an-odm-object-via-create_odm_from_r_matrix">Initializing an <code>odm</code> object via
<code>create_odm_from_r_matrix()</code><a class="anchor" aria-label="anchor" href="#initializing-an-odm-object-via-create_odm_from_r_matrix"></a>
</h2>
<p>We recommend that users create an <code>odm</code> object via
<code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code>, as this function is highly
scalable and typically requires only a couple gigabytes of memory.
However, users also can convert an R matrix into an <code>odm</code>
object via the function <code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>.
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> takes two main arguments:
<code>mat</code> and <code>file_to_write</code>. <code>mat</code> is a
standard R matrix (of type <code>"matrix"</code>) or a sparse R matrix
(of type <code>"dgCMatrix"</code>, <code>"dgRMatrix"</code>, or
<code>"dgTMatrix"</code>). <code>mat</code> should contain row names
giving the ID of each feature. Next, <code>file_to_write</code> is a
fully-qualified file path specifying the location in which to write the
backing <code>.odm</code> file. We provide an example of calling
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> on a gene-by-cell expression
matrix contained in the <code>sceptredata</code> package.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">lowmoi_example_data</span><span class="op">)</span></span>
<span><span class="va">gene_mat</span> <span class="op">&lt;-</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">response_matrix</span></span></code></pre></div>
<p><code>gene_mat</code> is a gene expression matrix containing 299
genes and 20,729 cells. (Users can evaluate
<code><a href="https://rdrr.io/pkg/sceptredata/man/lowmoi_example_data.html" class="external-link">?lowmoi_example_data</a></code> to see more information about this
matrix.) We pass this matrix to <code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code>,
setting <code>file_to_write</code> to
<code>paste0(temp_dir, "/gene.odm")</code>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_to_write</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"/gene.odm"</span><span class="op">)</span></span>
<span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">gene_mat</span>,</span>
<span>  file_to_write <span class="op">=</span> <span class="va">file_to_write</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>gene_odm</code> is a standard <code>odm</code> object.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_odm</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class <span style="color: #0000BB;">odm</span> with the following attributes:</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">299</span> features</span></span>
<span><span class="co">##  • <span style="color: #0000BB;">20729</span> cells</span></span>
<span><span class="co">##  • Backing file: <span style="color: #0000BB;">/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//Rtmp6grobL/gene.odm</span></span></span></code></pre>
<p>Moreover, the file <code>gene.odm</code> has been written to
<code>temp_dir</code>. (The previous <code>gene.odm</code> file is
overwritten.)</p>
</div>
<div class="section level2">
<h2 id="notes-on-compression">Notes on compression<a class="anchor" aria-label="anchor" href="#notes-on-compression"></a>
</h2>
<p><code><a href="../reference/create_odm_from_cellranger.html">create_odm_from_cellranger()</a></code> and
<code><a href="../reference/create_odm_from_r_matrix.html">create_odm_from_r_matrix()</a></code> take optional arguments
<code>chunk_size</code> and <code>compression_level</code> (which are
set to reasonable defaults). <code>chunk_size</code> and
<code>compression_level</code> control the extent to which the backing
<code>.odm</code> file is compressed. <code>chunk_size</code> should be
a positive integer, and <code>compression_level</code> should be an
integer in the range of 0 to 9. Increasing the value of these arguments
<em>increases</em> the level of compression, thereby leading to a
smaller file size for the backing <code>.odm</code> file (but possibly
longer read and write times).</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-lib/sessioninfo#readme" class="external-link">sessioninfo</a></span><span class="op">)</span>; <span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">##  version  R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">##  os       macOS Sonoma 14.3.1</span></span>
<span><span class="co">##  system   aarch64, darwin20</span></span>
<span><span class="co">##  ui       X11</span></span>
<span><span class="co">##  language en</span></span>
<span><span class="co">##  collate  en_US.UTF-8</span></span>
<span><span class="co">##  ctype    en_US.UTF-8</span></span>
<span><span class="co">##  tz       America/New_York</span></span>
<span><span class="co">##  date     2024-03-31</span></span>
<span><span class="co">##  pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">package    </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">##  BH            1.84.0-0 <span style="color: #555555;">2024-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  bit           4.0.5    <span style="color: #555555;">2022-11-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  bit64         4.0.5    <span style="color: #555555;">2020-08-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  bslib         0.7.0    <span style="color: #555555;">2024-03-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  cachem        1.0.8    <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  cli           3.6.2    <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  colorspace    2.1-0    <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  crayon        1.5.2    <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  data.table    1.15.4   <span style="color: #555555;">2024-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  desc          1.4.3    <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  digest        0.6.35   <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  evaluate      0.23     <span style="color: #555555;">2023-11-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  fansi         1.0.6    <span style="color: #555555;">2023-12-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  fastmap       1.1.1    <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  fs            1.6.3    <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  glue          1.7.0    <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  hms           1.1.3    <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  htmltools     0.5.8    <span style="color: #555555;">2024-03-25</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  jquerylib     0.1.4    <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  jsonlite      1.8.8    <span style="color: #555555;">2023-12-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  knitr         1.45     <span style="color: #555555;">2023-10-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  lattice       0.22-5   <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.3)</span></span></span>
<span><span class="co">##  lifecycle     1.0.4    <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  magrittr      2.0.3    <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  Matrix        1.6-5    <span style="color: #555555;">2024-01-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.3)</span></span></span>
<span><span class="co">##  memoise       2.0.1    <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  ondisc      * 1.1.0    <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  pillar        1.9.0    <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  pkgconfig     2.0.3    <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  pkgdown       2.0.7    <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  purrr         1.0.2    <span style="color: #555555;">2023-08-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  R.methodsS3   1.8.2    <span style="color: #555555;">2022-06-13</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  R.oo          1.26.0   <span style="color: #555555;">2024-01-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  R.utils       2.12.3   <span style="color: #555555;">2023-11-18</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  R6            2.5.1    <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  ragg          1.2.7    <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  Rcpp          1.0.12   <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  readr         2.1.5    <span style="color: #555555;">2024-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  Rhdf5lib      1.25.2   <span style="color: #555555;">2024-02-14</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  rlang         1.1.3    <span style="color: #555555;">2024-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  rmarkdown     2.26     <span style="color: #555555;">2024-03-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  rstudioapi    0.15.0   <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  sass          0.4.9    <span style="color: #555555;">2024-03-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  sceptre       0.9.1    <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  sceptredata * 0.9.0    <span style="color: #555555;">2024-03-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  sessioninfo * 1.2.2    <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  systemfonts   1.0.5    <span style="color: #555555;">2023-10-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  textshaping   0.3.7    <span style="color: #555555;">2023-10-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  tibble        3.2.1    <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  tidyselect    1.2.1    <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  tzdb          0.4.0    <span style="color: #555555;">2023-05-12</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  utf8          1.2.4    <span style="color: #555555;">2023-10-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  vctrs         0.6.5    <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  vroom         1.6.5    <span style="color: #555555;">2023-12-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  xfun          0.43     <span style="color: #555555;">2024-03-25</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  yaml          2.3.8    <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #555555;"> [1] /private/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T/RtmprqK5HM/temp_libpath1583c420cc4</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
