<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Single-cell datasets are growing in size, posing challenges as well as opportunities for genomics researchers. `ondisc` is an R package that facilitates analysis of large-scale single-cell data out-of-core on a laptop or distributed across tens to hundreds processors on a cluster or cloud. In both of these settings, `ondisc` requires only a few gigabytes of memory, even if the input data are tens of gigabytes in size. `ondisc` mainly is oriented toward single-cell CRISPR screen analysis, but ondisc also can be used for single-cell differential expression and single-cell co-expression analyses. ondisc is powered by several new, efficient algorithms for manipulating and querying large, sparse expression matrices."><title>ondisc: Algorithms and data structures for large single-cell expression matrices — ondisc-package • ondisc</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ondisc: Algorithms and data structures for large single-cell expression matrices — ondisc-package"><meta property="og:description" content="Single-cell datasets are growing in size, posing challenges as well as opportunities for genomics researchers. `ondisc` is an R package that facilitates analysis of large-scale single-cell data out-of-core on a laptop or distributed across tens to hundreds processors on a cluster or cloud. In both of these settings, `ondisc` requires only a few gigabytes of memory, even if the input data are tens of gigabytes in size. `ondisc` mainly is oriented toward single-cell CRISPR screen analysis, but ondisc also can be used for single-cell differential expression and single-cell co-expression analyses. ondisc is powered by several new, efficient algorithms for manipulating and querying large, sparse expression matrices."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ondisc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/ondisc.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/timothy-barry/ondisc/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>ondisc: Algorithms and data structures for large single-cell expression matrices</h1>
      <small class="dont-index">Source: <a href="https://github.com/timothy-barry/ondisc/blob/HEAD/R/ondisc.R" class="external-link"><code>R/ondisc.R</code></a></small>
      <div class="d-none name"><code>ondisc-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Single-cell datasets are growing in size, posing challenges as well as opportunities for genomics researchers. `ondisc` is an R package that facilitates analysis of large-scale single-cell data out-of-core on a laptop or distributed across tens to hundreds processors on a cluster or cloud. In both of these settings, `ondisc` requires only a few gigabytes of memory, even if the input data are tens of gigabytes in size. `ondisc` mainly is oriented toward single-cell CRISPR screen analysis, but ondisc also can be used for single-cell differential expression and single-cell co-expression analyses. ondisc is powered by several new, efficient algorithms for manipulating and querying large, sparse expression matrices.</p>
    </div>


    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://timothy-barry.github.io/ondisc/" class="external-link">https://timothy-barry.github.io/ondisc/</a></p></li>
<li><p><a href="https://timothy-barry.github.io/sceptre-book/" class="external-link">https://timothy-barry.github.io/sceptre-book/</a></p></li>
<li><p>Report bugs at <a href="https://github.com/timothy-barry/ondisc/issues" class="external-link">https://github.com/timothy-barry/ondisc/issues</a></p></li>
</ul></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><strong>Maintainer</strong>: Timothy Barry <a href="mailto:tbarry@hsph.harvard.edu">tbarry@hsph.harvard.edu</a> (<a href="https://orcid.org/0000-0002-4356-627X" class="external-link">ORCID</a>)</p>
<p>Authors:</p><ul><li><p>Eugene Katsevich <a href="mailto:ekatsevi@wharton.upenn.edu">ekatsevi@wharton.upenn.edu</a> [thesis advisor]</p></li>
</ul><p>Other contributors:</p><ul><li><p>Songcheng Dai [contributor]</p></li>
<li><p>Yixuan Qiu [contributor]</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># initialize odm objects from Cell Ranger output; also, compute the cellwise covariates</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/sceptre-book/" class="external-link">sceptredata</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">directories_to_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptredata"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">directory_to_write</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_odm_from_cellranger.html">create_odm_from_cellranger</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  directories_to_load <span class="op">=</span> <span class="va">directories_to_load</span>,</span></span>
<span class="r-in"><span>  directory_to_write <span class="op">=</span> <span class="va">directory_to_write</span>,</span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Round 1/2 processing of the input files.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Processing file 1 of 2.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Processing file 2 of 2.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Round 2/2 processing of the input files.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Processing file 1 of 2. Computing cellwise covariates. Writing to disk.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Processing file 2 of 2. Computing cellwise covariates. Writing to disk.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract the odm corresponding to the gene modality</span></span></span>
<span class="r-in"><span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="va">out_list</span><span class="op">$</span><span class="va">gene</span></span></span>
<span class="r-in"><span><span class="va">gene_odm</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class <span style="color: #0000BB;">odm</span> with the following attributes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• <span style="color: #0000BB;">526</span> features</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• <span style="color: #0000BB;">45919</span> cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• Backing file: <span style="color: #0000BB;">/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpQcfq9t/gene.odm</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># obtain dimension information</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]   526 45919</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 526</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 45919</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># obtain rownames (i.e., the feature IDs)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="rownames-odm-method.html">rownames</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ENSG00000069275" "ENSG00000117222" "ENSG00000117266" "ENSG00000117280"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "ENSG00000133059" "ENSG00000133065"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract row into memory, first by integer and then by string</span></span></span>
<span class="r-in"><span><span class="va">expression_vector_1</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="fl">10</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">expression_vector_2</span> <span class="op">&lt;-</span> <span class="va">gene_odm</span><span class="op">[</span><span class="st">"ENSG00000135046"</span>,<span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># delete the gene_odm object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">gene_odm</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># reinitialize the gene_odm object</span></span></span>
<span class="r-in"><span><span class="va">gene_odm</span> <span class="op">&lt;-</span> <span class="fu"><a href="initialize_odm_from_backing_file.html">initialize_odm_from_backing_file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/gene.odm"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_odm</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class <span style="color: #0000BB;">odm</span> with the following attributes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• <span style="color: #0000BB;">526</span> features</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• <span style="color: #0000BB;">45919</span> cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	• Backing file: <span style="color: #0000BB;">/var/folders/7v/5sqjgh8j28lgf8qx3gbtq1h00000gp/T//RtmpQcfq9t/gene.odm</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

